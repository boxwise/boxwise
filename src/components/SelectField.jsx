import React from "react";
import TextField from "@material-ui/core/TextField";

const SelectField = ({
  label,
  name,
  value,
  defaultText,
  items,
  itemToId,
  itemToText,
  dataTestId,
  ...props
}) => (
  <TextField
    label={label}
    name={name}
    id={name}
    select
    fullWidth
    margin="normal"
    value={value || ""}
    SelectProps={{
      // we use native select elements for a better mobile experience
      // plus it's easier to test
      native: true,
      inputProps: {
        "data-testid": dataTestId
      }
    }}
    {...props}
  >
    <option key="" value="">
      {defaultText}
    </option>
    {items.map(item => (
      <option key={itemToId(item)} value={itemToId(item)}>
        {itemToText(item)}
      </option>
    ))}
  </TextField>
);

export default SelectField;
