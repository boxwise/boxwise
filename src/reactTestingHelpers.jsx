import React from "react";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import { Provider } from "react-redux";
import { render, fireEvent } from "react-testing-library";
import { createMuiTheme, MuiThemeProvider } from "@material-ui/core/styles";
import thunk from "redux-thunk";
import { createStore, applyMiddleware } from "redux";

export { fireEvent, wait } from "react-testing-library";

// following suggestions in https://testing-library.com/docs/react-testing-library/setup#custom-render

export const typeText = (element, value) => {
  fireEvent.change(element, { target: { value } });
};

const theme = createMuiTheme({
  typography: {
    useNextVariants: true
  }
});

export function customRender(
  ui,
  {
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] }),
    initialState = {},
    reducer = state => state
  } = {}
) {
  const store = createStore(reducer, initialState, applyMiddleware(thunk));
  return render(
    <MuiThemeProvider theme={theme}>
      <Provider store={store}>
        <Router history={history}>{ui}</Router>
      </Provider>
    </MuiThemeProvider>
  );
}

export { customRender as render };
