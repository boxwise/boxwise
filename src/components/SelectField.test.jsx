import React from "react";

import { render, fireEvent } from "reactTestingHelpers";

import SelectField from "./SelectField";

describe("BoxList", () => {
  it("renders correctly with no items", () => {
    const { container } = render(
      <SelectField name="SomeField" label="Some Field" items={[]} />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly with items", () => {
    const { container } = render(
      <SelectField
        name="SomeField"
        label="Some Field"
        items={[{ id: "123", text: "Hello" }]}
        itemToId={({ id }) => id}
        itemToText={({ id, text }) => `${id}: ${text}`}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("triggers onChange", () => {
    const { container, getByTestId } = render(
      <SelectField
        name="SomeField"
        label="Some Field"
        items={[{ id: "123", text: "Hello" }]}
        itemToId={({ id }) => id}
        itemToText={({ id, text }) => `${id}: ${text}`}
        dataTestId="filter"
      />
    );
    fireEvent.change(getByTestId("filter"), { target: { value: "xyz" } });
    expect(container).toMatchSnapshot();
  });
});
