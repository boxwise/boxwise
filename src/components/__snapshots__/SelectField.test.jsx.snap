// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxList renders correctly with items 1`] = `
<div>
  <div
    class="MuiFormControl-root-1 MuiFormControl-marginDense-3 MuiFormControl-fullWidth-4"
  >
    <label
      class="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-marginDense-11"
      data-shrink="false"
    >
      Some Field
    </label>
    <div
      class="MuiInputBase-root-43 MuiInput-root-30 MuiInput-underline-34 MuiInputBase-fullWidth-52 MuiInput-fullWidth-37 MuiInputBase-formControl-44 MuiInput-formControl-31 MuiInputBase-marginDense-50"
    >
      <div
        class="MuiSelect-root-23"
      >
        <select
          aria-invalid="false"
          class="MuiSelect-select-24 MuiInputBase-input-53 MuiInput-input-38 MuiInputBase-inputMarginDense-54 MuiInput-inputMarginDense-39"
          name="SomeField"
        >
          <option
            value=""
          />
          <option
            value="123"
          >
            123: Hello
          </option>
        </select>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-60 MuiSelect-icon-29"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`BoxList renders correctly with no items 1`] = `
<div>
  <div
    class="MuiFormControl-root-1 MuiFormControl-marginDense-3 MuiFormControl-fullWidth-4"
  >
    <label
      class="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-marginDense-11"
      data-shrink="false"
    >
      Some Field
    </label>
    <div
      class="MuiInputBase-root-43 MuiInput-root-30 MuiInput-underline-34 MuiInputBase-fullWidth-52 MuiInput-fullWidth-37 MuiInputBase-formControl-44 MuiInput-formControl-31 MuiInputBase-marginDense-50"
    >
      <div
        class="MuiSelect-root-23"
      >
        <select
          aria-invalid="false"
          class="MuiSelect-select-24 MuiInputBase-input-53 MuiInput-input-38 MuiInputBase-inputMarginDense-54 MuiInput-inputMarginDense-39"
          name="SomeField"
        >
          <option
            value=""
          />
        </select>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-60 MuiSelect-icon-29"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`BoxList triggers onChange 1`] = `
<div>
  <div
    class="MuiFormControl-root-1 MuiFormControl-marginDense-3 MuiFormControl-fullWidth-4"
  >
    <label
      class="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-marginDense-11"
      data-shrink="false"
    >
      Some Field
    </label>
    <div
      class="MuiInputBase-root-43 MuiInput-root-30 MuiInput-underline-34 MuiInputBase-fullWidth-52 MuiInput-fullWidth-37 MuiInputBase-formControl-44 MuiInput-formControl-31 MuiInputBase-marginDense-50"
    >
      <div
        class="MuiSelect-root-23"
      >
        <select
          aria-invalid="false"
          class="MuiSelect-select-24 MuiInputBase-input-53 MuiInput-input-38 MuiInputBase-inputMarginDense-54 MuiInput-inputMarginDense-39"
          data-testid="filter"
          name="SomeField"
        >
          <option
            value=""
          />
          <option
            value="123"
          >
            123: Hello
          </option>
        </select>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-60 MuiSelect-icon-29"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
